engine/runtime-parser/src/main/java/org/enso/compiler/core/IR.java

```diff
@@ -71,16 +71,36 @@ public interface IR {
    */
   List<IR> children();
 
+  /**
+   * Applies the callback to nodes in the preorder walk of the tree of this node.
+   *
+   * @param ir the node to traverse
+   * @param cb the callback to apply
+   */
+  static void preorder(IR ir, Consumer<IR> cb) {
```

Comment:
I so much wish `Consumer` was a partial function.

Reply:
`scala.PartialFunction[IR, IR]` is nice. But implementing that on java side, not so much. You have to explicitly override both `isDefinedAt` and `apply` methods. Which is unnecessary duplication.

Reply:
We can make a variant for a ;PartialFunction; as well, but Pavel is right, it cannot be used as a lambda from Java.